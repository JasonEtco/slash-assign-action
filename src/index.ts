import { Toolkit } from 'actions-toolkit'
import slashAssignAction from './lib'

Toolkit.run(slashAssignAction, {
  event: ['issue_comment.created', 'schedule']
})
