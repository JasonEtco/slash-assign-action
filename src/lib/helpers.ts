import { SlashAssignToolkit } from "..";

// Have to explicitly get these, because `tools.inputs` is a Proxy`
export function getInputsFromTools (tools: SlashAssignToolkit) {
  return {
    assigned_label: tools.inputs.assigned_label,
    required_label: tools.inputs.required_label,
    pin_label: tools.inputs.pin_label,
    days_until_warning: tools.inputs.days_until_warning,
    days_until_unassign: tools.inputs.days_until_unassign,
    stale_assignment_label: tools.inputs.stale_assignment_label,
    assigned_comment: tools.inputs.assigned_comment,
  }
}