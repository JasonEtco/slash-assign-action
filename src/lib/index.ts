import { Toolkit } from 'actions-toolkit'
import commentHandler from './comment-handler'
import scheduleHandler from './schedule-handler'

export interface Inputs {
  assigned_label: string
  required_label?: string
  pin_label?: string
  days_until_warning: string
  days_until_unassign: string
  stale_assignment_label: string
  [key: string]: string | undefined
}

export type SlashAssignToolkit = Toolkit<Inputs>

export default async function slashAssignAction(tools: SlashAssignToolkit) {
  switch (tools.context.event) {
    case 'issue_comment':
      commentHandler(tools)
      break
    case 'schedule':
      await scheduleHandler(tools)
      break
    default:
      throw new Error(`Unhandled event ${tools.context.event}`)
  }
}
