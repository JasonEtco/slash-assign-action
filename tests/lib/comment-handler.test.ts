import nock from 'nock'
import { SlashAssignToolkit } from '../../src'
import commentHandler from '../../src/lib/comment-handler'
import { generateToolkit, recordNockRequests } from '../helpers'

describe('comment-handler', () => {
  let tools: SlashAssignToolkit

  beforeEach(() => {
    nock.cleanAll()
    tools = generateToolkit()
  })

  afterEach(() => {
    delete process.env.INPUT_REQUIRED_LABEL
  })

  it('assigns the user to the issue', async () => {
    const { scopedNock, requests } = recordNockRequests([
      {
        uri: '/repos/JasonEtco/testing/issues/1/assignees',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/labels',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/comments',
        method: 'post',
        response: { status: 200 }
      }
    ])

    await commentHandler(tools)
    expect(scopedNock.isDone()).toBe(true)
    expect(tools.exit.failure).not.toHaveBeenCalled()

    const [
      assignRequest,
      labelRequest,
      commentRequest
    ] = requests

    expect(assignRequest.params.assignees).toEqual([tools.context.payload.comment.user.login])
    expect(labelRequest.params.labels).toEqual(['slash-assigned'])
    expect(commentRequest.params.body).toMatchSnapshot()
  })

  it('exits early if the issue is already assigned', async () => {
    tools.context.payload.issue!.assignee = 'Saladin'
    await commentHandler(tools)
    expect(tools.exit.failure).toHaveBeenCalled()
    expect(tools.exit.failure).toHaveBeenCalledWith('Issue #1 is already assigned to @Saladin')
    delete tools.context.payload.issue?.assignee
  })

  it('exits early if required_label is set but not present', async () => {
    process.env.INPUT_REQUIRED_LABEL = 'required'
    await commentHandler(tools)
    expect(tools.exit.failure).toHaveBeenCalled()
    expect(tools.exit.failure).toHaveBeenCalledWith('Required label [required] label not found in issue #1.')
  })

  it('assigns the user if required_label is set and present', async () => {
    process.env.INPUT_REQUIRED_LABEL = 'required'
    tools.context.payload.issue!.labels = [{ name: 'required' }]

    const { scopedNock } = recordNockRequests([
      {
        uri: '/repos/JasonEtco/testing/issues/1/assignees',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/labels',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/comments',
        method: 'post',
        response: { status: 200 }
      }
    ])

    await commentHandler(tools)
    expect(tools.exit.failure).not.toHaveBeenCalled()
    expect(scopedNock.isDone()).toBe(true)
    tools.context.payload.issue!.labels = []
  })

  it('uses a custom assigned_comment message', async () => {
    process.env.INPUT_ASSIGNED_COMMENT = 'Assigned to @{{ comment.user.login }}'

    const { scopedNock, requests } = recordNockRequests([
      {
        uri: '/repos/JasonEtco/testing/issues/1/assignees',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/labels',
        method: 'post',
        response: { status: 200 }
      }, {
        uri: '/repos/JasonEtco/testing/issues/1/comments',
        method: 'post',
        response: { status: 200 }
      }
    ])

    await commentHandler(tools)
    expect(tools.exit.failure).not.toHaveBeenCalled()
    expect(scopedNock.isDone()).toBe(true)
    expect(requests[2].params.body).toBe('Assigned to @Shaxx')
  })
})
