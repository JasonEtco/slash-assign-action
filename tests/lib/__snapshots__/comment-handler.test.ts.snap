// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment-handler assigns the user to the issue 1`] = `
"This issue [has been assigned](https:&#x2F;&#x2F;github.com&#x2F;my-comment) to @Shaxx!

It will become unassigned if it isn't closed within 21 days. A maintainer can also add the **pinned** label to prevent it from being unassigned."
`;
